File: 2.1/python/main.py

two_count = 0
three_count = 0

with open('./../input.txt', 'r') as f:
    for line in f: 
        chars = {}
        for char in line:
            if char in chars:
                chars[char] += 1
            else: 
                chars[char] = 1
        
        two_found = False
        three_found = False
        for char, count in chars.items():
            if count == 2 and not two_found:
                two_count += 1
                two_found = True
            elif count == 3 and not three_found:
                three_count += 1
                three_found = True
            if two_found and three_found:
                break

print('Two count: ' + str(two_count))
print('Three count: ' + str(three_count))
print('Result checksum: ' + str(two_count*three_count))

File: 2.1/rust/src/main.rs
use std::fs::File;
use std::string::String;
use std::io::prelude::*;
use std::collections::HashMap;
use std::char;

fn main() {

    let mut box_ids = String::new();
    File::open("./../input.txt")
        .expect("Unable to read from input.txt")
        .read_to_string(&mut box_ids)
        .expect("Unable to feed contents to string");

    let mut two_letter_count = 0;
    let mut three_letter_count = 0;
    for box_id in box_ids.lines() {
        let mut character_count: HashMap<char, u8> = HashMap::new();
        for character in box_id.trim().chars() {
            let character_counter = character_count.entry(character).or_insert(0);
            *character_counter += 1;
        }

        let mut two_letter_found = false;
        let mut three_letter_found = false;
        for count in character_count.values() {
            if *count == 2 && !two_letter_found {
                two_letter_count += 1;
                two_letter_found = true;
            } else if *count == 3 && !three_letter_found {
                three_letter_count += 1;
                three_letter_found = true;
            }

            if two_letter_found && three_letter_found {
                break;
            }
        }
    }

    println!("Checksum: {}", two_letter_count * three_letter_count);
}
